src/libraries/Common/src/System/Reflection/Metadata/TypeNameParserHelpers.cs

```diff
@@ -0,0 +1,416 @@
+                    }
+                    else if (NeedsEscaping(c))
+                    {
+                        break;
+                    }
+                }
+                return offset;
+            }
+
+            static bool NeedsEscaping(char c) => c is '[' or ']' or '&' or '*' or ',' or '+' or EscapeCharacter;
+        }
+
+        internal static ReadOnlySpan<char> GetName(ReadOnlySpan<char> fullName)
+        {
+            int offset = fullName.LastIndexOfAny('.', '+');
+
+            if (offset > 0 && fullName[offset - 1] == EscapeCharacter) // this should be very rare (IL Emit & pure IL)
+            {
+                offset = GetUnescapedOffset(fullName, startIndex: offset);
+            }
+
+            return offset < 0 ? fullName : fullName.Slice(offset + 1);
+
+            static int GetUnescapedOffset(ReadOnlySpan<char> fullName, int startIndex)
+            {
+                int offset = startIndex;
+                for (; offset >= 0; offset--)
+                {
+                    if (fullName[offset] is '.' or '+')
+                    {
+                        if (offset == 0 || fullName[offset - 1] != EscapeCharacter)
+                        {
+                            break;
+                        }
+                        offset--; // skip the escaping character
+                    }
+                }
+                return offset;
+            }
+        }
+
+        internal static string GetRankOrModifierStringRepresentation(int rankOrModifier, ValueStringBuilder builder)
+        {
+            if (rankOrModifier == ByRef)
+            {
+                builder.Append('&');
+            }
+            else if (rankOrModifier == Pointer)
+            {
+                builder.Append('*');
+            }
+            else if (rankOrModifier == SZArray)
+            {
+                builder.Append("[]");
+            }
+            else if (rankOrModifier == 1)
+            {
+                builder.Append("[*]");
+            }
+            else
+            {
+                Debug.Assert(rankOrModifier >= 2 && rankOrModifier <= 32);
+
+                builder.Append('[');
+                for (int i = 1; i < rankOrModifier; i++)
+                {
+                    builder.Append(',');
+                }
+                builder.Append(']');
+            }
+
+            return builder.ToString();
+        }
+
+        /// <summary>
+        /// Are there any captured generic args? We'll look for "[[" and "[" that is not followed by "]", "*" and ",".
+        /// </summary>
+        internal static bool IsBeginningOfGenericAgs(ref ReadOnlySpan<char> span, out bool doubleBrackets)
+        {
+            doubleBrackets = false;
+
+            if (!span.IsEmpty && span[0] == '[')
+            {
+                // There are no spaces allowed before the first '[', but spaces are allowed after that.
+                ReadOnlySpan<char> trimmed = span.Slice(1).TrimStart();
```

Comment:
The `Trim*` APIs consume more than the space (`U+0020`) character. They consume every character [listed here](https://util.unicode.org/UnicodeJsps/list-unicodeset.jsp?a=[:White_Space=Yes:]). Is this intentional and desirable?

(This comment applies to all calls to `Trim*` throughout this PR.)

Reply:
I am using `TrimStart` everywhere because the CLR parser was so far using `char.IsWhiteSpace`.

https://github.com/dotnet/runtime/blob/3725bfc33df7e65a6c86c9341d078a1db6ff5ab1/src/libraries/Common/src/System/Reflection/TypeNameParser.cs#L448-L450